### ISSUE TITLE
Issues in Safari on iOS 

### ISSUE THREAD

**Issue by volkanunsal** [heroku/react-refetch#54]
It looks like there are a couple of bugs in iOS. The first one is that a polyfill is required for `Object.assign`. Secondly, the PromiseState fails to resolve in iOS. I'm using the latest release candidate.

**Comment by ryanbrainard** [comment]
@volkanunsal 

> It looks like there are a couple of bugs in iOS.

Do you mean in React Native? This has not been tested at all with React Native. I would love someone who is more familiar (or has a real use case) in that space to try out, doc, and maybe even improve compatibility.  

> The first one is that a polyfill is required for Object.assign. 

Correct me if I'm wrong, but that should probably be the responsibility of the app. Would documentation around this help?

> PromiseState fails to resolve in iOS

Any ideas why? Reproducers?

**Comment by volkanunsal** [comment]
No, I mean the Safari browser on iPhone. I'm testing in an emulator.

> Correct me if I'm wrong, but that should probably be the responsibility of the app. 

Mentioning that we need to add polyfills for some ES6 features in the documentation would be helpful. It seems like the headline for it would be the library doesn't support IE and Safari by default. 

My guess on the PromiseState issue is it probably also requires a polyfill for a Promise. Or maybe the implementation of the Promise object in Safari is faulty. I haven't tested this idea yet, but I will report what comes out of it.

**Comment by volkanunsal** [comment]
I checked the Promise issue. The interesting thing about it is that one of the promises are fulfilled, and the second one isn't. This only happens on Safari. Other browsers fulfill both requests. I will try to get a test case up, but my code looks like this:

```javascript
@refetch(props => {
  let fetchers = {
    dirFetch: {
      url: `https://api.github.com/repos/volkanunsal/docs/contents/pages`,
      then: (pages) => ({value: pages})
    }
  };
  if (props.params.splat) {
    fetchers.pathFetch = {
      url: `https://api.github.com/repos/volkanunsal/docs/contents/${props.params.splat}`,
      then: (page) => ({ value: page.content ? atob(page.content) : '' })
    };
  }

  return fetchers;
})
export default class DocsPage extends React.Component {
  render() {
    const { dirFetch, pathFetch } = this.props;
    let pages = [];
    console.log(dirFetch);
    // => Fulfilled
    console.log(pathFetch);
    // => Not fulfilled
  }
}
```

**Comment by volkanunsal** [comment]
Oh you're also using Request! That's not supported by Safari either, and it's marked as [experimental technology in Javascript reference docs.](https://developer.mozilla.org/en-US/docs/Web/API/Request)

**Comment by volkanunsal** [comment]
It seems like the Request object is polyfilled by [whatwg-fetch library.](https://github.com/github/fetch/blob/a9f49ab0875e284dbd0eda9be5004949e31b4161/fetch.js#L215) But it's important to realize that library is [only included if the native Promise object doesn't exist.](https://github.com/github/fetch/blob/a9f49ab0875e284dbd0eda9be5004949e31b4161/fetch.js#L384) Meaning browsers that support Promise, but not Request, like Safari, won't be able to use it.

**Comment by volkanunsal** [comment]
It looks like the problem was using `atob` in the `then` function. This probably took longer than it should in Safari and caused some problems with how Promises get fulfilled. Now the promise gets fulfilled, but it won't return the correct value. It returns an empty string. Again only on Safari.

**Comment by volkanunsal** [comment]
Looks like this is an issue with [Safari's implementation of `atob`](http://stackoverflow.com/a/27536313/32816). I guess that takes care of the second problem.

**Comment by ryanbrainard** [comment]
@volkanunsal Thanks digging in and investigating this. As far as action items:

 - Document recommended polyfills (note, there is also a [plan](https://github.com/heroku/react-refetch/issues/11) to remove `whatwg-fetch` as a peer dependency and make it just an app selectable polyfill, as needed)
 - Document this issue for Safari
 - Anything else for this project?

Note, as far as `Request` being experiment, the whole [Fetch API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API) is considered experimental.

**Comment by volkanunsal** [comment]
@ryanbrainard Awesome. Yeah, I think that's all of it.

**Comment by ryanbrainard** [comment]
Re-opening as a reminder for the action items above.

**Comment by ryanbrainard** [comment]
Closed by #76 and release in [v1.0.0-beta.0](https://github.com/heroku/react-refetch/releases/tag/v1.0.0-beta.0)
